hedera-node/hedera-app/src/main/java/com/hedera/node/app/signature/hapi/SignatureVerificationFuture.java

```diff
@@ -0,0 +1,305 @@
+                final var txSig = map.get(key);
+                yield txSig != null && txSig.getSignatureStatus() == VerificationStatus.VALID;
+            }
+            case ECDSA_384, RSA_3072, CONTRACT_ID, DELEGATABLE_CONTRACT_ID, UNSET -> false;
+            case KEY_LIST -> checkIfPassed(key.keyListOrThrow(), map);
+            case THRESHOLD_KEY -> checkIfPassed(key.thresholdKeyOrThrow(), map);
+        };
+    }
+
+    /**
+     * Checks whether the given {@link ThresholdKey} has been verified. This kind of key only requires a subset of
+     * the keys to be valid. The subset that are not valid may not be valid because they are just the wrong kind of
+     * key (for example, a contract ID key), or because there was corresponding signature in the signature map for
+     * the key, or because there was a key but it failed key verification.
+     *
+     * @param thresholdKey The {@link ThresholdKey} to verify. If the threshold is less than or equal to zero, then
+     *                     the threshold is clamped to 1. If the threshold is greater than the number of keys, then
+     *                     it is clamped to the number of keys. There must be at least one key to be valid.
+     * @param map A Map of {@link Key} to {@link TransactionSignature}
+     * @return {@code true} if at least {@link ThresholdKey#threshold()} keys within the {@code key} have been
+     * validated, {@code false} otherwise.
+     */
+    private boolean checkIfPassed(
+            @NonNull final ThresholdKey thresholdKey, @NonNull final Map<Key, TransactionSignature> map) {
+        if (thresholdKey.hasKeys()) {
+            final var keyList = thresholdKey.keysOrThrow();
+            if (keyList.hasKeys()) {
+                final var list = keyList.keysOrThrow();
+                if (list.isEmpty()) return false;
+                final var threshold = Math.max(1, Math.min(list.size(), thresholdKey.threshold()));
+                var numCanFail = list.size() - threshold;
+                for (final var subKey : list) {
+                    final var passed = checkIfPassed(subKey, map);
+                    if (!passed && --numCanFail < 0) {
+                        return false;
+                    }
+                }
+                return true;
+            }
+        }
+        return false;
+    }
+
+    /**
+     * Checks whether the given {@link KeyList} has been verified. Such a key is valid if, and only if, every key
+     * within the {@link KeyList} has been validated.
+     *
+     * @param keyList The {@link KeyList} to verify. Must have at least one key to be valid.
+     * @param map A Map of {@link Key} to {@link TransactionSignature}
+     * @return {@code true} if at least {@link ThresholdKey#threshold()} keys within the {@code key} have been
+     * validated, {@code false} otherwise.
+     */
+    private boolean checkIfPassed(@NonNull final KeyList keyList, @NonNull final Map<Key, TransactionSignature> map) {
```

Comment:
There is a special case that if we are checking the signature for a file's `KeyList` during `FileDelete`, we need to pass this check if _any_ key in the list signs (c.f. [here](https://github.com/hashgraph/hedera-protobufs/blob/main/services/file_delete.proto#L32)).

Reply:
I saw that comment, but it seems very surprising (and wrong!). Why do we do that? One disgruntled signer can nuke a file?

Reply:
Maybe what the `FileDelete` will need to do is shred the keylists. Basically, given a `k1`, create a new key `k2`, such that all key lists are converted into threshold keys with a threshold of 1.